src/py21cmfast/_memory.py

```diff
@@ -0,0 +1,706 @@
+    *,
+    user_params=None,
+):
+    """A function to estimate total memory usage of a brightness_temperature call."""
+    """Memory usage of Python BrightnessTemp class."""
+
+    """All declared HII_DIM boxes"""
+    # brightness_temp
+    num_py_boxes = 1.0
+
+    size_py = num_py_boxes * (user_params.HII_DIM) ** 3
+
+    # These are all float arrays
+    size_py = (np.float32(1.0).nbytes) * size_py
+
+    """Memory usage within BrightnessTemperatureBox.c"""
+    hii_tot_fft_num_pixels = (
+        2.0 * (float(user_params.HII_DIM) / 2.0 + 1.0) * (user_params.HII_DIM) ** 2
+    )
+
+    # box, unfiltered_box
+    num_c_boxes = 2.0
+
+    size_c = (np.float32(1.0).nbytes) * num_c_boxes * hii_tot_fft_num_pixels
+
+    return {"python": size_py, "c": size_c}
+
+
+def mem_halo_field(
+    *,
+    user_params=None,
+):
+    """A function to estimate total memory usage of a determine_halo_list call."""
+    """Memory usage of Python HaloField class."""
+
+    """All declared DIM boxes"""
+    # halo_field
+    num_py_boxes = 1.0
+
+    size_py = num_py_boxes * (user_params.DIM) ** 3
+
+    # These are all float arrays
+    size_py = (np.float32(1.0).nbytes) * size_py
+
+    """Memory usage within FindHaloes.c"""
+    kspace_num_pixels = (float(user_params.DIM) / 2.0 + 1.0) * (user_params.DIM) ** 2
+
+    # density_field, density_field_saved
+    num_c_boxes = 2.0
+
+    # halo_field, in_halo
+    num_c_boxes_alt = (
+        1.25  # in_halo is size 1 (char *) so count it as 0.25 of float (4)
+    )
+
+    if global_params.OPTIMIZE:
+        # forbidden
+        num_c_boxes_alt += 0.25
+
+    # These are fftwf complex arrays (thus 2 * size)
+    size_c = (2.0 * np.float32(1.0).nbytes) * num_c_boxes * kspace_num_pixels
+
+    size_c += (np.float32(1.0).nbytes) * num_c_boxes_alt * (user_params.DIM) ** 3
+
+    # We don't know a priori how many haloes that will be found, but we'll estimate the memory usage
+    # at 10 per cent of the total number of pixels (HII_DIM, likely an over estimate)
+    # Below the factor of 4 corresponds to the mass and three spatial locations. It is defined as an
+    # int but I'll leave it as 4 bytes in case
+    size_c += 0.1 * 4.0 * (np.float32(1.0).nbytes) * (user_params.HII_DIM) ** 3
+
+    return {"python": size_py, "c": size_c}
+
+
+def mem_perturb_halo(
+    *,
+    user_params=None,
+):
+    """A function to estimate total memory usage of a perturb_halo_list call."""
+    """Memory usage of Python PerturbHaloField class."""
+
+    # We don't know a priori how many haloes that will be found, but we'll estimate the memory usage
+    # at 10 per cent of the total number of pixels (HII_DIM, likely an over estimate)
+    # Below the factor of 4 corresponds to the mass and three spatial locations. It is defined as an
+    # int but I'll leave it as 4 bytes in case
+    size_c = 0.1 * 4.0 * (np.float32(1.0).nbytes) * (user_params.HII_DIM) ** 3
+
+    return {"python": 0.0, "c": size_c}
+
+
+def format_output(
```

Comment:
Yep, I threw this together in a few minutes at the end of the day. I tried to use the inbuilt logging rather than printing but it didn't work (hence the printing). But I didn't try too hard to fix the issue